### ISSUE TITLE
&dyn Display not match type of interger literal without explicit annotated

### ISSUE THREAD

**Issue by occurrent** [rust-lang/rust-analyzer#13150]
see example below

``` rust
use std::{fmt::{Display, Formatter, Result}};

trait D{
    // copy from std lib
    fn fmt(&self, _: &mut Formatter<'_>) -> Result;
}

impl D for i32{
    fn fmt(&self, _: &mut Formatter<'_>) -> Result{
        todo!()
    }
}

fn main() {
    let i1 = 1;
    let i2 = 1i32;
    let i3: i32 = 1;

    // ra: error
    let _ : &dyn Display = &i1;
    // ra: pass
    let _ : &dyn Display = &i2;
    let _ : &dyn Display = &i3;

    // ra: error
    let _ : &dyn Display = &1;
    let _ : &dyn Display = &1.;

    // ra: pass
    let _ : &dyn D = &1;

    // ra: pass
    let _ : &dyn Display = &"";
    let _ : &dyn Display = &1i32;
    let _ : &dyn Display = &1f64;
}
```
The above code was reported as an error by **ra**, but passed by `cargo check` and `rustc`.

**rust-analyzer version:** 0.3.1186-standalone (e8e598f64 2022-08-28)
**rustc version:** rustc 1.63.0 (4b91a6ea7 2022-08-08)

**Comment by ChayimFriedman2** [comment]
Dup #11847.
